File: DesignPatternsInAutomatedTesting-Series/PatternsInAutomation.Tests/Advanced/Decorator/Advanced/Strategies/OrderPurchaseStrategyDecorator.cs
using System;
using PatternsInAutomatedTests.Advanced.Decorator.Data;

namespace PatternsInAutomatedTests.Advanced.Decorator.Advanced.Strategies
{
    public abstract class OrderPurchaseStrategyDecorator : OrderPurchaseStrategy
    {
        protected readonly OrderPurchaseStrategy orderPurchaseStrategy;
        protected readonly ClientPurchaseInfo clientPurchaseInfo;
        protected readonly decimal itemsPrice;

        public OrderPurchaseStrategyDecorator(OrderPurchaseStrategy orderPurchaseStrategy, decimal itemsPrice, ClientPurchaseInfo clientPurchaseInfo)
        {
            this.orderPurchaseStrategy = orderPurchaseStrategy;
            this.itemsPrice = itemsPrice;
            this.clientPurchaseInfo = clientPurchaseInfo;
        }

        public override decimal CalculateTotalPrice()
        {
            this.ValidateOrderStrategy();

            return this.orderPurchaseStrategy.CalculateTotalPrice();
        }

        public override void ValidateOrderSummary(decimal totalPrice)
        {
            this.ValidateOrderStrategy();
            this.orderPurchaseStrategy.ValidateOrderSummary(totalPrice);
        }

        private void ValidateOrderStrategy()
        {
            if (this.orderPurchaseStrategy == null)
            {
                throw new Exception("The OrderPurchaseStrategy should be first initialized.");
            }
        }
    }
}

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/decorator/advanced/strategies/OrderPurchaseStrategyDecorator.java
package decorator.advanced.strategies;

import decorator.data.PurchaseInfo;

import java.math.BigDecimal;

public class OrderPurchaseStrategyDecorator extends OrderPurchaseStrategy {
    protected final OrderPurchaseStrategy orderPurchaseStrategy;
    protected final PurchaseInfo purchaseInfo;
    protected final BigDecimal itemPrice;

    public OrderPurchaseStrategyDecorator(OrderPurchaseStrategy orderPurchaseStrategy, BigDecimal itemPrice, PurchaseInfo purchaseInfo) {
        this.orderPurchaseStrategy = orderPurchaseStrategy;
        this.itemPrice = itemPrice;
        this.purchaseInfo = purchaseInfo;
    }

    @Override
    public BigDecimal calculateTotalPrice() {
        validateOrderStrategy();
        return orderPurchaseStrategy.calculateTotalPrice();
    }

    @Override
    public void validateOrderSummary(BigDecimal totalPrice) {
        validateOrderStrategy();
        orderPurchaseStrategy.validateOrderSummary(totalPrice);
    }

    private void validateOrderStrategy() {
        if (orderPurchaseStrategy == null) {
            throw new NullPointerException("The OrderPurchaseStrategy should be first initialized.");
        }
    }
}
